addons/sale_coupon/models/coupon_program.py

```diff
@@ -175,6 +175,11 @@ def _filter_programs_from_common_rules(self, order, next_order=False):
             programs += programs_curr_order._filter_not_ordered_reward_programs(order)
         return programs
 
+    def _get_discount_product_values(self):
+        res = super()._get_discount_product_values()
+        res['invoice_policy'] = 'order'
+        return res
+
```

Comment:
IMHO, the products created in coupon could already be created with the `invoice_policy` set to `order`.
Otherwise, the products programs created before installation of sale_coupon would still be invoiced on delivery by default.

Unless I'm missing something in the POS flow, I believe there won't be any problem if we always create the discount products with the "Ordered Qties" invoicing policy.

Reply:
I don't think pos will be affected on setting the discount product to automatically have `invoice_policy = 'order'`. I believe this can be put in the core coupon module. Besides, 'order' is the default 'invoice_policy' of a created product, no?

Reply:
Not anymore, since https://github.com/odoo/odoo/pull/47526 (14.0), but it will probably be reverted soon in master (there is an ongoing task to bring back the 'order' invoice_policy by default).

Reply:
I see. But it's fine for pos to have the discount product to be 'order discount policy'.

Reply:
Good then, let's change the behavior in the base `coupon` module then :+1:.
Thanks for the confirmation :).

Reply:
@Feyensv ,
When the sale_coupon is installed and default invoicing is set for `delivered Quantities` in Sale config then on creation of coupon program new reward line(product) will be created and it's default policy is `delivered Quantities` instead of `ordered Quantities` so to avoid this we added the invoice_policy `order` on creation of reward line(product)

Thanks

Reply:
Yes, I understand the problem, I just meant that we can specify the invoice_policy directly in coupon, no need to add a specific override in `sale_coupon`.  The products for reward/promotion programs should always be created with the `order` `invoice_policy`, even if `sale` is not installed.

Reply:
@Feyensv , And yes when only coupon is installed then there is no `invoice_policy` field as it comes from the `sale` module.  Thanks

Reply:
Ow, my bad indeed :man_facepalming:.
Then this fix is perfect :+1:
(Forget what I said before :D)